mne/beamformer/_dics.py

```diff
@@ -1,113 +1,440 @@
+
+            if pick_ori == 'max-power':
+                # Compute spectral power by applying the spatial filters to the
+                # CSD matrix.
+                power = Wk.dot(Cm).dot(Wk.T)
+
+                # Compute the direction of max power
+                u, s, _ = np.linalg.svd(power)
+                max_power_ori = u[:, 0].real
+                max_power_ori /= np.linalg.norm(max_power_ori)
+
+                # Re-compute the filter in the direction of max power
+                Gk_max = Gk.dot(max_power_ori)
+                Wk_max = max_power_ori.dot(Wk)
+                Ck_max = np.dot(Wk_max, Gk_max)
+                Wk_max /= Ck_max
+                Wk[:] = Wk_max
+
+        if pick_ori == 'normal':
+            W = W[2::3]
+        elif pick_ori == 'max-power':
+            W = W[0::3]
+
+        Ws.append(W)
+    Ws = np.array(Ws)
 
-        # Noise normalization
-        noise_norm = np.dot(np.dot(Wk.conj(), noise_csd.data), Wk.T)
-        noise_norm = np.abs(noise_norm).trace()
-        Wk /= np.sqrt(noise_norm)
+    subject = _subject_from_forward(forward)
+    filters = dict(weights=Ws, csd=csd, ch_names=ch_names, proj=proj,
+                   vertices=vertices, subject=subject,
+                   n_orient=n_orient if pick_ori is None else 1)
+
+    return filters
 
-    # Pick source orientation normal to cortical surface
-    if pick_ori == 'normal':
-        W = W[2::3]
-        is_free_ori = False
 
+def _apply_dics(data, filters, info, tmin):
+    """Apply DICS spatial filter to data for source reconstruction."""
     if isinstance(data, np.ndarray) and data.ndim == 2:
         data = [data]
-        return_single = True
+        one_epoch = True
     else:
-        return_single = False
+        one_epoch = False
 
-    subject = _subject_from_forward(forward)
+    Ws = filters['weights']
+    one_freq = len(Ws) == 1
+
+    subject = filters['subject']
     for i, M in enumerate(data):
-        if len(M) != len(picks):
+        if len(M) != len(filters['ch_names']):
             raise ValueError('data and picks must have the same length')
 
-        if not return_single:
+        if not one_epoch:
             logger.info("Processing epoch : %d" % (i + 1))
 
         # Apply SSPs
         if info['projs']:
-            M = np.dot(proj, M)
+            _check_proj_match(info, filters)
+            M = np.dot(filters['proj'], M)
 
-        # project to source space using beamformer weights
-        if is_free_ori:
+        stcs = []
+        for W in Ws:
+            # project to source space using beamformer weights
             sol = np.dot(W, M)
-            logger.info('combining the current components...')
-            sol = combine_xyz(sol)
+
+            if filters['n_orient'] > 1:
+                logger.info('combining the current components...')
+                sol = combine_xyz(sol)
+
+            tstep = 1.0 / info['sfreq']
+
+            # XXX : STC cannot contain (yet?) complex values
```

Comment:
maybe we should save to HDF5 in this case

Reply:
Now that I think about it, power estimates should always be real numbers, otherwise you cannot interpret them. If any conversion needs to take place, converting it before putting it into an STC object is as good a place as any to do it. For the application of DICS to raw timecourses (epochs, evokeds), we may need the complex result for further computations.

Reply:
In which case, yes, it would make sense to store as HDF5.